### PR TITLE
Improve nvm detector: locate installed nvm, report it path, version, installed node.js versions

### PR DESCRIPTION
**Pull request by VladRassokhin**

### CODE REVIEW

**Comment by jonnyzzz** [comment]
Could you please rebase it on master. It looks we have 2 PR here. One for NVM option (I like it!) and the other to improve the detector.

**Comment by VladRassokhin** [comment]
This PR split into two.
I'm thinking of renaming `node.js.nvm.installed.versions` to `node.js.installed.versions` and further add support for [nvm-windows](https://github.com/coreybutler/nvm-windows) or other tools.
Also RubyEnvironmentConfigurator-like build feature could be created that would support nvm, nvm-windows, etc.

**Comment by VladRassokhin** [comment]
Intermediate results:

Agent properties:
```text
node.js	 4.4.5
node.js.npm	 2.15.5
node.js.nvm.dir	 /home/vagrant/.nvm
node.js.nvm.installable	 yes
node.js.nvm.installed.versions	 v4.4.5,v6.2.2
node.js.nvm.version	 0.31.1
``` 

This properties refreshed after each build step, as it may install some tools
Problem: seems 'node.js' should be '6.2.2' in this case. Same for 'node.js.npm'

File: agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt [hunk 1]
```diff
@@ -40,29 +37,43 @@ class NodeToolsDetector(events: EventDispatcher<AgentLifeCycleListener>,
...
           }
         }
       }
     }
+
+    val nvm_dir = locateInstalledNVM()
+    if (nvm_dir != null) {
+      val version = getNVMVersion(nvm_dir) ?: "N/A"
+      LOG.info("NVM installation detected at '$nvm_dir' with version '$version'")
+      val versions = getNVMInstalledNodeVersions(nvm_dir)?.joinToString(",").orEmpty()
```
Comment (jonnyzzz) [inline agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt]: please consider `", "` as separator, not `","`

File: agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt [hunk 2]
```diff
@@ -40,29 +37,43 @@ class NodeToolsDetector(events: EventDispatcher<AgentLifeCycleListener>,
...
             addConfigurationParameter(NPMBean().nodeJSNPMConfigurationParameter, ref)
             addConfigurationParameter(NodeBean().nodeJSConfigurationParameter, ref)
-            addConfigurationParameter(NVMBean().NVMAvailable, "yes")
+            addConfigurationParameter(NVMBean().NVMInstallable, "yes")
           }
         }
       }
     }
+
+    val nvm_dir = locateInstalledNVM()
```
Comment (jonnyzzz) [inline agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt]: No logging for the case `locateInstalledNVM() == null`

File: agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt [hunk 3]
```diff
@@ -108,4 +130,53 @@ class NodeToolsDetector(events: EventDispatcher<AgentLifeCycleListener>,
       }
     })
   }
+
+  private fun setVerbose(verbose: Boolean) {
+    // TODO: either modify logger or add some boolean field which used before all LOG calls
```
Comment (jonnyzzz) [inline agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt]: This looks like a hack

File: agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt [hunk 4]
```diff
@@ -74,15 +85,26 @@ class NodeToolsDetector(events: EventDispatcher<AgentLifeCycleListener>,
...
+        if (!run.stdOut.isEmptyOrSpaces()) LOG.info("StdOut: ${run.stdOut}")
+        if (!run.stdErr.isEmptyOrSpaces()) LOG.info("StdErr: ${run.stdErr}")
       }
     }
   }
 
   init {
     events.addListener(object : AgentLifeCycleAdapter() {
       override fun beforeAgentConfigurationLoaded(agent: BuildAgent) {
+        setVerbose(true)
```
Comment (jonnyzzz) [inline agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt]: It's better to have `doDetectTools` have a parameter to specify the way logging is performed. 
I do not like an idea to reconfigure logger in any way somewhere here

File: agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt [hunk 5]
```diff
@@ -108,4 +130,53 @@ class NodeToolsDetector(events: EventDispatcher<AgentLifeCycleListener>,
...
+    val nvm_home = System.getenv("NVM_HOME")
+    if (nvm_home != null && isNVMInstallation(File(nvm_home))) return nvm_home
+
+    val home = System.getProperty("user.home")
+    if (home != null) {
+      val candidate = File(home, ".nvm")
+      if (isNVMInstallation(candidate)) return candidate.absolutePath
+    }
+
+    return null
```
Comment (jonnyzzz) [inline agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt]: Why not to have a configuration parameter in `buildAgent.properties` to override default search path?

Comment (jonnyzzz) [inline agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt]: Here goes no logging why it failed to find NVM

File: agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt [hunk 6]
```diff
@@ -108,4 +130,53 @@ class NodeToolsDetector(events: EventDispatcher<AgentLifeCycleListener>,
...
+  }
+
+  private fun getNVMInstalledNodeVersions(nvm_dir: String): List<String>? {
+    // Use next command (checked with nvm 0.7.0, 0.31.1) instead of 'nvm ls' as it returns simple list without any colors
+    // bash -c ". $NVM_DIR/nvm.sh; echo \"\$(nvm_ls)\""
+    val run = runNVMAware(nvm_dir, "echo \"$(nvm_ls)\"")
+    if (!run.succeeded()) {
+      LOG.warn("Failed to detect installed nvm node.js versions: ${run.stdErr}")
```
Comment (jonnyzzz) [inline agent/src/com/jonnyzzz/teamcity/plugins/node/agent/NodeDetector.kt]: exit code?

**Comment by jonnyzzz** [comment]
Could you please create tests. You may comeup with a `Dockerfile` to create necessary environment to check if all commands work

**Comment by srizzling** [comment]
Is this something that could be picked up?

Our usecase for a feature like this, is that we can prebuild node versions into a agent (installed via nvm) and use this plugin to switch between, instead of wrapping ugly source ~/.nvm/nvm.sh commands, meaning we don't need to download each version each time to neglect network affects

**Pull request closed without merging**